### PR TITLE
drainer support kprobe

### PR DESCRIPTION
**Pull request by runzexia**
Signed-off-by: runzexia <root@junwuhui.cn>
ref https://github.com/knative/eventing/issues/3911
https://github.com/knative/networking/blob/bf808e4cd3f37ea514d365ffb2bb5ffa908e141c/pkg/probe/handler.go#L47-L48

### CODE REVIEW

File: network/handlers/drain.go [hunk 1]
```diff
@@ -83,6 +84,19 @@ func (d *Drainer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		if d.draining() {
+			http.Error(w, "shutting down", http.StatusServiceUnavailable)
+		} else {
+			hh := r.Header.Get(network.HashHeaderName)
+			if hh == "" {
+				http.Error(w, fmt.Sprintf("a probe request must contain a non-empty %q header", network.HashHeaderName), http.StatusBadRequest)
+				return
+			}
+			w.Header().Set(network.HashHeaderName, hh)
+			w.WriteHeader(http.StatusOK)
```
Comment (mattmoor) [inline network/handlers/drain.go:97]: Can we factor a function for this?  We'll want to deduplicate this logic (incl. constants) when it lands in knative/networking, so we should make this reusable.

**Comment by mattmoor** [comment]
cc @ZhiminXiang @tcnghia @nak3

File: network/handlers/drain.go (hunk 1 above)
Comment (runzexia) [reply network/handlers/drain.go]: done😁

File: network/handlers/drain_test.go [hunk 2]
```diff
@@ -236,6 +236,58 @@ func TestDrainMechanics(t *testing.T) {
...
 }
 
+func TestDrainerKProbe(t *testing.T) {
+	var (
+		w          http.ResponseWriter
+		req        = &http.Request{}
+		kprobehash = "hash"
+		kprobe     = &http.Request{
+			Header: http.Header{
+				"K-Network-Probe": []string{"probe"},
```
Comment (nak3) [inline network/handlers/drain_test.go:246]: Could you use the const value `K-Network-Probe`, `K-Network-Hash` and `probe` defined in network/network.go?

File: network/network_test.go [hunk 3]
```diff
@@ -42,3 +43,58 @@ func TestIsKubeletProbe(t *testing.T) {
...
+
+func TestIsKnativeProbe(t *testing.T) {
+	req, err := http.NewRequest(http.MethodGet, "http://example.com/", nil)
+	if err != nil {
+		t.Fatal("Error building request:", err)
+	}
+	if IsKProbe(req) {
+		t.Error("Not a knative probe but counted as such")
+	}
+	req.Header.Set("K-Network-Probe", "probe")
```
Comment (nak3) [inline network/network_test.go:55]: Same with above. Could you use the const values in this test file?

File: network/network_test.go [hunk 4]
```diff
@@ -42,3 +43,58 @@ func TestIsKubeletProbe(t *testing.T) {
...
+	if !IsKProbe(req) {
+		t.Error("knative probe but not counted as such")
+	}
+	req.Header.Del("K-Network-Probe")
+	if IsKProbe(req) {
+		t.Error("Not a knative probe but counted as such")
+	}
+	req.Header.Set("K-Network-Probe", "no matter")
+	if IsKProbe(req) {
+		t.Error("Not a knative probe but not counted as such")
```
Comment (nak3) [inline network/network_test.go:65]: This test would be `"Not a knative probe but counted as such"`?

**Review by nak3 — CHANGES_REQUESTED** [review]

File: network/handlers/drain_test.go (hunk 2 above)
Comment (runzexia) [reply network/handlers/drain_test.go]: ok, thanks for review😋

**Review by nak3 — APPROVED** [review]

**Review by mattmoor — COMMENTED** [review]
/lgtm

**Review by tcnghia — APPROVED** [review]

**Pull request merged**